Add Sphinx documentation (#6)
metricq/metricq-sink-nsca: docs/usage/configuration.rst

                        "critical_below": 5.0
                    }
                }
            }

:code:`reporting_host`
    Name of the host for which check results are reported, as configured in Nagios/Centreon.

    Default:
        The current hostname returned by :manpage:`gethostname(2)`.

:code:`resend_interval`
    .. _global_resend_interval:

    Global default resend interval (see :ref:`Check configuration/resend_interval<resend_interval>`).

    Default
        :literal:`"3min"`

Check configuration
^^^^^^^^^^^^^^^^^^^

A single check monitors a set of metrics for abnormal behaviour.
It continuously consumes new data points for these metrics
and reports and `overall state`: if values of a single metric exceed their :ref:`allowed range<abnormal-ranges>`
or there are no new values after :ref:`a certain time<timeout>`
a state of :literal:`WARNING` or :literal:`CRITICAL` is reported.

:code:`metrics` (list of strings)
    A list of metrics that should be monitored.

    This list is **mandatory** and required to be *non-empty*.

:code:`warning_above`, :code:`warning_below`, :code:`critical_above`, :code:`critical_below` (number)
    .. _abnormal-ranges:

> 

Reviewer: Mention that setting at least one of these values mean that the nsca-sink needs to decode and process the packets, which adds significant overhead for metrics with large messages.

    Range of value which should trigger a :literal:`WARNING` (resp. :literal:`CRITICAL`) status report to be sent.
    We call the intervals :math:`[-∞, \mathtt{warning\_below}) \cup (\mathtt{warning\_above}, ∞]` the *warning range*;
    values within that range trigger a :literal:`WARNING`.
    The *critical range* is defined similarly.

>     The critical range must be contained within the warning range, otherwise the configuration is rejected.

Reviewer: ```suggestion
    The warning range must be contained within the critical range, otherwise, the configuration is rejected.
```

Author: Hm... this is technically correct, since the range of values for which warnings are generated contains all values for which `criticals` are generated. But you are right, this sounds confusing.

Reviewer: Oh, so I misread "range"...  As this sentence is confusing and redundant to the provided term below. I'd suggest to just leave it out.

    Explicitly, we required

    .. math::
        \mathtt{critical\_below}
            ≤ \mathtt{warning\_below}
            < \mathtt{warning\_above}
            ≤ \mathtt{critical\_above}

    Defaults

>         * :math:`-∞` (:code:`{warning,critical}_below`)

Reviewer: Emphasize that the default values mean that the nsca sink won't performe any value checks.

        * :math:`∞`  (:code:`{warning,critical}_above`)

        You cannot put :math:`±∞` directly into the check configuration.
        Since they are the default anyway, simply omit the relevant key if necessary.

    Example
        An example check configuration with all ranges specified:

        .. code-block:: json

            {
                "checks": {
                    "foo": {
                        "metrics": [ "foo.bar", "foo.baz" ],
                        "critical_below":   5.0,
                        "warning_below":   10.0,
                        "warning_above":   95.0,
                        "critical_above": 100.0
                    }
                }
            }

:code:`timeout` (:ref:`duration<Duration>`)
    .. _timeout:

>     Send check result of severity :literal:`WARNING` if values arrive apart more than the specified period.

Reviewer: Emphasize that the timeout checks alone don't require processing of the messages and thus, can be enabled for metrics with large messages or high sampling rate.

    .. note::

        **Use with care**.
        The implementation essentially only performs a floating-point equality test to filter values.

        If this sounds like a bad idea to you, you are probably right.
        Trust me, this is here because some source cannot be fixed easily.

    Example
        A source computing a `Power factor <https://en.wikipedia.org/wiki/Power_factor>`_
        for an AC electrical power system reports a metric :literal:`ac_system.power_factor`.
        If the power factor is too low, a warning should be generated.
        It might be that the source calculates a power factor of :literal:`0.0` on low draw.
        Since a low power factor on low draw might not be considered a problem,
        ignore the value :literal:`0.0`:

        .. code-block:: json

            {
                "checks": {
                    "low-draw": {
                        "metrics": ["ac_system.power_factor"],
                        "warning_below": 0.8,
                        "ignore": [0.0]
                    }
                }
            }

:code:`resend_interval` (:ref:`duration<Duration>`)
    .. _resend_interval:

>     Period of time after which the current state of this check is sent again to the server, eventhough it might not have changed.

Reviewer: ```suggestion
    Period of time after which the current state of this check is sent again to the server, even though it might not have changed.
```

    This is necessary since passive checks are considered to be in an :literal:`UNKNOWN` state by Centreon/Nagios
    if they have not sent a report for a certain time.

    Default
        Inherited from the :ref:`global resend interval<global_resend_interval>`.

:code:`transition_debounce_window` (:ref:`duration<Duration>`)
    If necessary, :command:`metricq_sink_nsca` tries to reduce the number of spurious :literal:`WARNING` or :literal:`CRITICAL` reports.
    We call this process *"transition debouncing"*.
    It does so by keeping a history of state transitions for each metric,

>     only sending an abnormal report if the majority of recent state transitions indicate an abnormal state.

Reviewer: I don't understand what it does.

Provide an example and/or improve the description.

        :literal:`"localhost"`

:code:`port` (`integer`)
    Port of the NSCA daemon to which check results are sent.
    See :code:`-p` flag of :code:`send_nsca`.

    Default:
        :literal:`5667`

:code:`executable` (`string`)
    Path to :code:`send_nsca` executable to use for sending check results.

    Default:
        :literal:`"/usr/sbin/send_nsca"`

:code:`config_file` (`string`)
    Path to :code:`send_nsca` configuration file.
    See :code:`-c` flag of :code:`send_nsca`

    Default:
        :literal:`"/etc/nsca/send_nsca.cfg"`

Plugin configuration
^^^^^^^^^^^^^^^^^^^^

:code:`file` (string)
    File system path to plugin implementation (a `.py` file).

    This key is **mandatory**.
    Plugin configurations without it are rejected.

:code:`config` (dictionary)
    An arbitrary JSON object containing a plugin-specific configuration.

Duration
^^^^^^^^

All durations in this configuration are *strings* in the form of :samp:`"{value} [{unit}]"`.
The :samp:`{value}` is an integer or (decimal) floating point literal, the unit is one of

* :literal:`s`/:literal:`seconds`
* :literal:`ms`/:literal:`milliseconds`
* :literal:`us`/:literal:`μs`/:literal:`microseconds`,

> * :literal:`ns`/:literal:`nanoseconds`

Reviewer: ~~Catch~~ Sort 'em all!

* :literal:`d`/:literal:`days`
* :literal:`h`/:literal:`hours`
* :literal:`min`/:literal:`minutes`

If the unit is not specified, the value is interpreted as a *number of seconds*.

Examples
    * :literal:`"5s"`, :literal:`"5"` (5 seconds)
    * :literal:`"42 milliseconds"`
    * :literal:`"1.5 days"`

Complete Example
----------------

> An example with all possible options set is given below:

Reviewer: The example looks a bit borked. `plugins` should be part of the root object, don't they? It's also missing `resend_interval` and `reporting_host` in the root node.

Author: No, `plugins` are configured per-check.

But I'll add the other missing items to the example configuration.

    Default
        :literal:`"3min"`

Check configuration
^^^^^^^^^^^^^^^^^^^

A single check monitors a set of metrics for abnormal behavior.
It continuously consumes new data points for these metrics
and reports and `overall state`: if values of a single metric exceed their :ref:`allowed range<abnormal-ranges>`
or there are no new values after :ref:`a certain time<timeout>`
a state of :literal:`WARNING` or :literal:`CRITICAL` is reported.

:code:`metrics` (list of strings)
    A list of metrics that should be monitored.

    This list is **mandatory** and required to be *non-empty*.

:code:`warning_above`, :code:`warning_below`, :code:`critical_above`, :code:`critical_below` (number)
    .. _abnormal-ranges:

    Range of value which should trigger a :literal:`WARNING` (resp. :literal:`CRITICAL`) status report to be sent.
    We call the intervals :math:`[-∞, \mathtt{warning\_below}) \cup (\mathtt{warning\_above}, ∞]` the *warning range*;
    values within that range trigger a :literal:`WARNING`.
    The *critical range* is defined similarly.

    The critical range must be contained within the warning range, otherwise the configuration is rejected.

>     Explicitly, we required

Reviewer: ```suggestion
    Explicitly, we require
```

Author: Done here:

https://github.com/metricq/metricq-sink-nsca/pull/6/commits/5384c03f492225ee3c1f8d61cce4c63ebeaffc66#diff-563b65cda960bf6375d0d1218f42a8b9003545c5decdcbd18de9b2cbe073b346R110
